pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/custom/TimestampIndexTest.java

```diff
@@ -0,0 +1,183 @@
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.apache.pinot.util.TestUtils;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+
+public class TimestampIndexTest extends BaseClusterIntegrationTest implements ExplainIntegrationTestTrait {
+  @BeforeClass
+  public void setUp()
+      throws Exception {
+    TestUtils.ensureDirectoriesExistAndEmpty(_tempDir, _segmentDir, _tarDir);
+
+    // Start the Pinot cluster
+    startZk();
+    startController();
+    startBroker();
+    startServers(2);
+
+    // Create and upload the schema and table config
+    Schema schema = createSchema();
+    addSchema(schema);
+    TableConfig tableConfig = createOfflineTableConfig();
+    addTableConfig(tableConfig);
+
+    // Unpack the Avro files
+    List<File> avroFiles = unpackAvroData(_tempDir);
+
+    // Create and upload segments
+    ClusterIntegrationTestUtils.buildSegmentsFromAvro(avroFiles, tableConfig, schema, 0, _segmentDir, _tarDir);
+    uploadSegments(getTableName(), _tarDir);
+
+    // Wait for all documents loaded
+    waitForAllDocsLoaded(600_000L);
+  }
+
+  protected void overrideBrokerConf(PinotConfiguration brokerConf) {
+    String property = CommonConstants.MultiStageQueryRunner.KEY_OF_MULTISTAGE_EXPLAIN_INCLUDE_SEGMENT_PLAN;
+    brokerConf.setProperty(property, "true");
+  }
+
+  @Test
+  public void timestampIndexSubstitutedInProjectionsMSE() {
+    setUseMultiStageQueryEngine(true);
+    explain("SELECT datetrunc('SECOND',ArrTime) FROM mytable",
+        "Execution Plan\n"
+            + "PinotLogicalExchange(distribution=[broadcast])\n"
+            + "  LeafStageCombineOperator(table=[mytable])\n"
+            + "    StreamingInstanceResponse\n"
+            + "      StreamingCombineSelect\n"
+            + "        SelectStreaming(table=[mytable], totalDocs=[115545])\n"
+            + "          Project(columns=[[$ArrTime$SECOND]])\n"
+            + "            DocIdSet(maxDocs=[120000])\n"
+            + "              FilterMatchEntireSegment(numDocs=[115545])\n");
+  }
+
+  @Test
+  public void timestampIndexSubstitutedInFiltersMSE() {
+    setUseMultiStageQueryEngine(true);
+    explain("SELECT 1 FROM mytable where datetrunc('SECOND',ArrTime) > 1",
+        "Execution Plan\n"
+            + "PinotLogicalExchange(distribution=[broadcast])\n"
+            + "  LeafStageCombineOperator(table=[mytable])\n"
+            + "    StreamingInstanceResponse\n"
+            + "      StreamingCombineSelect\n"
+            + "        SelectStreaming(table=[mytable], totalDocs=[115545])\n"
+            + "          Transform(expressions=[['1']])\n"
+            + "            Project(columns=[[]])\n"
+            + "              DocIdSet(maxDocs=[120000])\n"
+            + "                FilterRangeIndex(predicate=[$ArrTime$SECOND > '1'], "
+            + "indexLookUp=[range_index], operator=[RANGE])\n");
+  }
+
```

Comment:
Would it be worth checking queries with :
- GROUP BY datetrunc('SECOND',ArrTime) 
- HAVING datetrunc('SECOND',ArrTime) > 1
- HAVING sum(case when datetrunc('SECOND',ArrTime) > 1 then 2 else 0 end) > 1  
?

Reply:
> GROUP BY datetrunc('SECOND',ArrTime)

Added

> HAVING datetrunc('SECOND',ArrTime) > 1
> HAVING sum(case when datetrunc('SECOND',ArrTime) > 1 then 2 else 0 end) > 1

In these cases the optimization is not supported in SSE and it isn't easy to support it. The main problem here is that `InstancePlanMakerImplV2.rewriteQueryContextWithHints` rewrites the query using a dirty trick: it modifies the lists returned by `queryContext.getSelectExpressions()`, `queryContext.getGroupByExpressions()`, etc. In the `queryContext.getAggregationFunctions()` case, it returns a list of `AggregationFunction` which have their own `AggregationFunction.getInputExpressions()`. But it is not guaranteed that modifying the list returned by that method actually modifies the `AggregationFunction`. In fact in cases most cases it doesn't.

> UNION-ed second query ?

Unions don't have expressions right? In case one of the unioned-queries have expressions, it should be treated by the other cases using recursion.

This tests class is testing more than this feature. It is testing that the substitution is being done, which means that the overrideHint is added (what this PR changes) and that the override is actually applied (what is failing in some cases). We could create a test that just verifies that the overrideHint is added, but that may be a complex test to add, so I think the tests we have right now are good enough.

Reply:
> WHERE EXISTS (select * from tab t2 where datetrunc('SECOND',ArrTime) > 1 )
> WHERE EXISTS (select * from tab t2 where t1.x = datetrunc('SECOND',t2.ArrTime) > 1 ) )

This is tricky. Would we want to test all possible SQL constructors? In theory... maybe... But I don't think it is worth it. We know we are applying the hint with a visitor that walks through the relational nodes. Therefore we know we will visit the sub-query filters. Which means that it is the same case as a select in the main query.

Obviously that is a grey area. A future refactor may not use a visitor and therefore may end up not visiting the inner query. But I don't think it is worth to test so deep. Probably, as said before, a lower level test (ie a unit test on the visitor) should make it more explicit why that case isn't worth to be tested.

Reply:
I checked HAVING on a few queries locally and think it's fine because HAVING expression is applied after aggregation by matching expression to group by output (so no additional aggregates are computed if it matches group by element).